util/src/main/java/io/kubernetes/client/util/Cache.java

```diff
@@ -0,0 +1,269 @@
+  }
+
+  private void init() throws ApiException, IOException {
+    if (list == null) {
+      updateList();
+    }
+    if (watch == null) {
+      startWatch();
+    }
+  }
+
+  public synchronized ApiListType list() throws ApiException, IOException {
+    init();
+    return list;
+  }
+
+  public synchronized List<ApiType> listForNamespace(String namespace)
+      throws ApiException, IOException {
+    init();
+    return namespaceIndex.get(namespace);
+  }
+
+  public synchronized ApiType get(String namespace, String name) throws ApiException, IOException {
+    init();
+    return getItem(namespace, name);
+  }
+
+  public void close() throws IOException {
+    if (watch != null) {
+      watch.close();
+    }
+  }
+
+  private Call makeCall(Boolean watch) {
+    CallGeneratorParams params = new CallGeneratorParams(watch, null, timeout);
+    return callGenerator.apply(params);
+  }
+
+  private void updateList() throws ApiException, IOException {
+    Call call = makeCall(Boolean.FALSE);
+    com.squareup.okhttp.Response response = call.execute();
+    if (!response.isSuccessful()) {
+      String respBody = null;
+      try (ResponseBody body = response.body()) {
+        if (body != null) {
+          respBody = response.body().string();
+        }
+      } catch (IOException e) {
+        throw new ApiException(
+            response.message(), e, response.code(), response.headers().toMultimap());
+      }
+      throw new ApiException(
+          response.message(), response.code(), response.headers().toMultimap(), respBody);
+    }
+    list =
+        (ApiListType) client.getJSON().getGson().fromJson(response.body().charStream(), listClass);
+    try {
+      List<ApiType> itemList = getItemList();
+      for (ApiType item : itemList) {
+        indexItem(item);
+      }
+    } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException ex) {
+      throw new ApiException(ex);
+    }
+  }
+
+  private ApiType getItem(String namespace, String name) {
+    try {
+      Method itemsField = listClass.getDeclaredMethod("getItems");
+      Object itemsObj = itemsField.invoke(list);
+      List<ApiType> itemList = (List<ApiType>) itemsObj;
+      for (ApiType obj : itemList) {
+        V1ObjectMeta meta = Reflect.objectMetadata(obj);
+        if (meta.getNamespace().equals(namespace) && meta.getName().equals(name)) {
+          return obj;
+        }
+      }
+    } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException ex) {
+      log.error("Error adding item", ex);
+      ex.printStackTrace();
+    }
+    return null;
+  }
+
+  private boolean sameObject(V1ObjectMeta a, V1ObjectMeta b) {
+    if (!b.getName().equals(a.getName())) {
+      return false;
+    }
+    if (b.getNamespace() == null && a.getNamespace() == null) {
+      return true;
+    }
+    return b.getNamespace().equals(a.getNamespace());
+  }
+
+  private int findItem(List<ApiType> itemList, ApiType item)
```

Comment:
There's some substantial dupliaction of code between this and `getItem` above? refactor and combine?

Reply:
combined them into reflection util class

https://github.com/kubernetes-client/java/pull/461/files#diff-958c385e8e8081adb4b361aa17de8756R35